File: MVVM/Visual Basic/MVVM/Helpers/Converters.vb
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.Globalization
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Data
Imports System.Windows.Media
Imports System.Windows.Navigation
Imports devDept.Eyeshot
Imports devDept.Graphics

Public Class ColorTableToStringConverter
    Implements IValueConverter
    Public Function IValueConverter_Convert(value As Object, targetType As Type, parameter As Object, culture As CultureInfo) As Object Implements IValueConverter.Convert
        Dim retVal As String = String.Empty
        If TypeOf value Is ObservableCollection(Of Brush) Then
            Select Case DirectCast(value, ObservableCollection(Of Brush)).Count
                Case 9
                    If True Then
                        retVal = "Red to Blue 9"
                        Exit Select
                    End If
                Case 17
                    If True Then
                        retVal = "Red to Blue 17"
                        Exit Select
                    End If
                Case 33
                    If True Then
                        retVal = "Red to Blue 33"
                        Exit Select
                    End If
            End Select
        End If
        Return retVal
    End Function
        
    Public Function IValueConverter_ConvertBack(value As Object, targetType As Type, parameter As Object, culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
        Throw New NotSupportedException()
    End Function
End Class

Public Class ViewportEnumsToStringConverter
    Implements IValueConverter
    Public Function IValueConverter_Convert(value As Object, targetType As Type, parameter As Object, culture As CultureInfo) As Object Implements IValueConverter.Convert
        If TypeOf value Is backgroundStyleType Then
            Return DirectCast(value, backgroundStyleType).ToString()
        End If

        If TypeOf value Is coordinateSystemPositionType Then
            Return DirectCast(value, coordinateSystemPositionType).ToString()
        End If

        If TypeOf value Is originSymbolStyleType Then
            Return DirectCast(value, originSymbolStyleType).ToString()
        End If

        If TypeOf value Is ToolBar.positionType Then
            Return DirectCast(value, ToolBar.positionType).ToString()
        End If

        If TypeOf value Is displayType Then
            Return DirectCast(value, displayType).ToString()
        End If

        If TypeOf value Is actionType Then
            Return DirectCast(value, actionType).ToString()
        End If

        If TypeOf value Is colorThemeType Then
            Return DirectCast(value, colorThemeType).ToString()
        End If

        Return [String].Empty
    End Function

    Public Function IValueConverter_ConvertBack(value As Object, targetType As Type, parameter As Object, culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
        Throw New NotSupportedException()
    End Function
End Class

<ValueConversion(GetType(Boolean), GetType(Boolean))> _
Public Class InverseBooleanConverter
	Implements IValueConverter
	Public Function IValueConverter_Convert(value As Object, targetType As Type, parameter As Object, culture As CultureInfo) As Object Implements IValueConverter.Convert
		If targetType <> GetType(Boolean) Then
			Throw New InvalidOperationException("The target must be a boolean")
		End If

		Return Not CBool(value)
	End Function

	Public Function IValueConverter_ConvertBack(value As Object, targetType As Type, parameter As Object, culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
		Throw New NotSupportedException()
	End Function
End Class

Public Class ColorToBrushConverter
    Implements IValueConverter
    Public Function IValueConverter_Convert(value As Object, targetType As Type, parameter As Object, culture As CultureInfo) As Object Implements IValueConverter.Convert
        If TypeOf value Is System.Windows.Media.Color Then
            Return Helper.ConvertColor(DirectCast(value, System.Windows.Media.Color))
        End If
        Return Nothing
    End Function

    Public Function IValueConverter_ConvertBack(value As Object, targetType As Type, parameter As Object, culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
        Throw New NotSupportedException()
    End Function
End Class

Public Class ColorToDrawingColorConverter
    Implements IValueConverter
    Public Function IValueConverter_Convert(value As Object, targetType As Type, parameter As Object, culture As CultureInfo) As Object Implements IValueConverter.Convert
        If TypeOf value Is System.Windows.Media.Color Then
            Return Helper.ConvertDrawingColor(DirectCast(value, System.Windows.Media.Color))
        End If
        Return Nothing
    End Function

    Public Function IValueConverter_ConvertBack(value As Object, targetType As Type, parameter As Object, culture As CultureInfo) As Object Implements IValueConverter.ConvertBack
        Throw New NotSupportedException()
    End Function
End Class

Public NotInheritable Class Helper
    Private Sub New()
    End Sub
    Public Shared Function ConvertColor(brush As Brush) As Color
        Dim newBrush As SolidColorBrush = DirectCast(brush, SolidColorBrush)
        Return newBrush.Color
    End Function

    Public Shared Function ConvertColor(color As Color) As Brush
        Return New SolidColorBrush(color)
    End Function

    Public Shared Function ConvertDrawingColor(color As Color) As System.Drawing.Color
        Return System.Drawing.Color.FromArgb(color.A, color.R, color.G, color.B)
    End Function

    ''' <summary>
    ''' Converts hex color string to <see cref="System.Drawing.Color"/>
    ''' </summary>
    ''' <param name="hexColor">Hex color like "#FF434752"</param>
    ''' <returns>The <see cref="System.Drawing.Color"/></returns>
    Public Shared Function ConvertColor(hexColor As String) As System.Drawing.Color
	    Return System.Drawing.ColorTranslator.FromHtml(hexColor)
    End Function

    Public Shared Function GetUriFromResource(resourceFilename As String) As Uri
        Return New Uri(Convert.ToString("pack://application:,,,/Resources/") & resourceFilename)
    End Function
End Class

File: MVVM/CSharp/MVVM/Helpers/Converters.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Data;
using System.Windows.Media;
using System.Windows.Navigation;
using devDept.Eyeshot;
using devDept.Graphics;

namespace WpfApplication1
{
    public class ColorTableToStringConverter : IValueConverter
    {
        public object Convert(object value, Type targetType, object parameter, System.Globalization.CultureInfo culture)
        {
            string retVal = string.Empty;
            if (value is ObservableCollection<Brush>)
            {
                switch (((ObservableCollection<Brush>)value).Count)
                {
                    case 9:
                        {
                            retVal = "Red to Blue 9";
                            break;
                        }
                    case 17:
                        {
                            retVal = "Red to Blue 17";
                            break;
                        }
                    case 33:
                        {
                            retVal = "Red to Blue 33";
                            break;
                        }
                }
            }
            return retVal;
        }

        public object ConvertBack(object value, Type targetType, object parameter, System.Globalization.CultureInfo culture)
        {
            throw new NotSupportedException();
        }
    }        

    public class ViewportEnumsToStringConverter : IValueConverter
    {
        public object Convert(object value, Type targetType, object parameter, System.Globalization.CultureInfo culture)
        {
            if (value is backgroundStyleType)
            {
                return ((backgroundStyleType)value).ToString();
            }

            if (value is coordinateSystemPositionType)
            {
                return ((coordinateSystemPositionType)value).ToString();
            }

            if (value is originSymbolStyleType)
            {
                return ((originSymbolStyleType)value).ToString();
            }

            if (value is ToolBar.positionType)
            {
                return ((ToolBar.positionType) value).ToString();
            }

            if (value is displayType)
            {
                return ((displayType) value).ToString();
            }

            if (value is actionType)
            {
                return ((actionType)value).ToString();
            }

            if (value is colorThemeType)
            {
                return ((colorThemeType)value).ToString();
            }

            return String.Empty;
        }

        public object ConvertBack(object value, Type targetType, object parameter, System.Globalization.CultureInfo culture)
        {
            throw new NotSupportedException();
        }
    }

    [ValueConversion(typeof(bool), typeof(bool))]
    public class InverseBooleanConverter : IValueConverter
    {        
        public object Convert(object value, Type targetType, object parameter,
            System.Globalization.CultureInfo culture)
        {
            if (targetType != typeof(bool))
                throw new InvalidOperationException("The target must be a boolean");

            return !(bool)value;
        }

        public object ConvertBack(object value, Type targetType, object parameter,
            System.Globalization.CultureInfo culture)
        {
            throw new NotSupportedException();
        }     
    }

    public class ColorToBrushConverter : IValueConverter
    {
        public object Convert(object value, Type targetType, object parameter, System.Globalization.CultureInfo culture)
        {
            if (value is System.Windows.Media.Color)
            {
                return Helper.ConvertColor((System.Windows.Media.Color)value);
            }
            return null;
        }

        public object ConvertBack(object value, Type targetType, object parameter, System.Globalization.CultureInfo culture)
        {
            throw new NotSupportedException();
        }
    }

    public class ColorToDrawingColorConverter : IValueConverter
    {
        public object Convert(object value, Type targetType, object parameter, System.Globalization.CultureInfo culture)
        {
            if (value is System.Windows.Media.Color)
            {
                return Helper.ConvertDrawingColor((System.Windows.Media.Color)value);
            }
            return null;
        }

        public object ConvertBack(object value, Type targetType, object parameter, System.Globalization.CultureInfo culture)
        {
            throw new NotSupportedException();
        }
    }

    public static class Helper
    {
        public static Color ConvertColor(Brush brush)
        {
            SolidColorBrush newBrush = (SolidColorBrush)brush;
            return newBrush.Color;
        }

        public static Brush ConvertColor(Color color)
        {
            return new SolidColorBrush(color);
        }

        public static System.Drawing.Color ConvertDrawingColor(Color color)
        {
            return System.Drawing.Color.FromArgb(color.A, color.R, color.G, color.B);
        }

        /// <summary>
        /// Converts hex color string to <see cref="System.Drawing.Color"/>
        /// </summary>
        /// <param name="hexColor">Hex color like "#FF434752"</param>
        /// <returns>The <see cref="System.Drawing.Color"/></returns>
        public static System.Drawing.Color ConvertColor(string hexColor)
        {            
            return System.Drawing.ColorTranslator.FromHtml(hexColor);
        }

        public static Uri GetUriFromResource(string resourceFilename)
        {
            return new Uri(@"pack://application:,,,/Resources/" + resourceFilename);
        }
    }
}
